src/binder/Binding.ts

```diff
@@ -0,0 +1,65 @@
+import { Key } from '../Key';
+import { Provider } from '../provider';
+import { Scope } from './ScopedBinder';
+
+export class Binding {
+  protected key: string;
+  protected isPromise: boolean;
+  protected provider?: Provider;
+  protected scope?: Scope;
+  protected resolved?: Provider;
+  protected dependencies?: Array<Key>;
+
+  constructor(key: string, asPromise = false) {
+    this.key = key;
+    this.isPromise = asPromise;
+  }
+
+  getKey(): string {
+    return this.key;
+  }
+
+  setProvider<Value>(provider: Provider<Value>): void {
+    this.provider = provider;
+  }
+
+  getProvider(): Binding['provider'] {
+    return this.provider;
+  }
+
+  setAsPromise(asPromise: boolean): void {
+    this.isPromise = asPromise;
+  }
+
+  asPromise(): boolean {
+    return this.isPromise;
+  }
+
+  setDependencies(keys: Array<Key>): void {
+    this.dependencies = keys;
+  }
+
+  hasDependencies(): boolean {
+    return this.dependencies !== undefined && this.dependencies.length > 0;
```

Comment:
```suggestion
    return this.dependencies?.length > 0;
```

Reply:
No, that would lead to a possible comparison of `undefined > 0`. I don't know if it works, but if it works it's more confusing than the two separate steps.

Reply:
falsey values will always evaluate to false when comparing. but I guess it's ok if we keep the two steps